File: src/main/java/net/kkolyan/astar/Finder.java
package net.kkolyan.astar;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.List;

public class Finder<P> {
	private Area<P> field;
	private NodeManager<P> nodeManager = new DefaultNodeManager<P>();
	private Collection<P> destinationPoints;
	private P sourcePoint;
	private List<P> path;
	private boolean done;

	//===========================================================================

	public void setDestinationPoints(Collection<P> destinationPoints) {
		this.destinationPoints = destinationPoints;
	}

	public void setSourcePoint(P sourcePoint) {
		this.sourcePoint = sourcePoint;
	}

	public void setField(Area<P> field) {
		this.field = field;
	}

	//===========================================================================

	/**
	 *
	 * @return list of points. excludes start and includes finish. null if there no path.
	 */
	public List<P> getPath() {
		return path;
	}

	//===========================================================================

	public void doSearch() {
		Node<P> node = new Node<P>();
		node.point = sourcePoint;
		node.costG = 0;
		node.costH = field.getHeuristicCost(node.point, destinationPoints);
		node.parent = null;
		nodeManager.openNode(node);
		while (!done) {
			searchAroundBest();
		}
	}

	//===========================================================================

	private void searchAroundBest() {
		Node<P> currentNode = nodeManager.getBestOpened();
		if (currentNode == null) {
			done = true;
			return;
		}
		nodeManager.closeNode(currentNode);

		for (P neighborPoint: field.getNeighbours(currentNode.point)) {
			if (destinationPoints.contains(neighborPoint)) {
				constructPath(currentNode, neighborPoint);
				done = true;
				return;
			}
			NodeState<P> info = nodeManager.getNodeState(neighborPoint);
			if (info != null) {
				if (info.isOpened() && currentNode.costG < info.getNode().parent.costG) {
					info.getNode().parent = currentNode;
				}
			} else {
				Node<P> neighbor = new Node<P>();
				neighbor.parent = currentNode;
				neighbor.point = neighborPoint;
				neighbor.costG = currentNode.costG + field.getStepCost(currentNode.point, neighborPoint);
				neighbor.costH = field.getHeuristicCost(neighbor.point, destinationPoints);
				nodeManager.openNode(neighbor);
			}
		}
	}

	//===========================================================================

	private void constructPath(Node<P> parent, P destination) {
		List<P> list = new ArrayList<P>();
        list.add(destination);
		while (parent.parent != null) {
			list.add(parent.point);
			parent = parent.parent;
		}
		Collections.reverse(list);
		path = list;
	}

	//===========================================================================
}

File: src/main/javascript/Finder.js

function Finder(data) {
    var source = data["source"];
    var destinations = data["destinations"];
    var area = data["area"];
    
    var nodeManager = new NodeManager();
    var destinationsHash = {};
    
    for (var i = 0; i < destinations.length; i ++) {
        destinationsHash[destinations[i].toIdString()] = true;
    }

    this.getNodeManager = function () {
        return nodeManager;
    };

    this.doSearch = function() {
        var node = {};
        node.point = source;
        node.costG = 0;
        node.costH = area.getHeuristicCost(node.point, destinations);
        node.parent = null;
        nodeManager.openNode(node);
        while (!this.done) {
            this.searchAroundBest();
        }
    };

    this.searchAroundBest = function() {
        var currentNode = nodeManager.getBestOpened();
        if (currentNode == null) {
            this.done = true;
            return;
        }
        nodeManager.closeNode(currentNode);

        var neighbors = area.getNeighbours(currentNode.point);
        for (var i in neighbors) {
            var neighborPoint = neighbors[i];
            if (destinationsHash[neighborPoint.toIdString()]) {
                this.constructPath(currentNode);
                this.done = true;
                return;
            }
            var nodeState = nodeManager.getNodeState(neighborPoint);
            if (nodeState != null) {
                if (nodeState.opened && currentNode.costG < nodeState.node.parent.costG) {
                    nodeState.node.parent = currentNode;
                }
            } else {
                var neighbor = {};
                neighbor.parent = currentNode;
                neighbor.point = neighborPoint;
                neighbor.costG = currentNode.costG + area.getStepCost(currentNode.point, neighborPoint);
                neighbor.costH = area.getHeuristicCost(neighbor.point, destinations);
                nodeManager.openNode(neighbor);
            }
        }
    };

    this.constructPath = function(parent) {
        var reversedList = [];
        while (parent.parent != null) {
            reversedList.push(parent.point);
            parent = parent.parent;
        }
        reversedList.reverse();
        this.path = reversedList;
    };
}
